src/vs/workbench/contrib/debug/browser/debugEditorContribution.ts

```diff
@@ -722,39 +770,18 @@ export class DebugEditorContribution implements IDebugEditorContribution {
 				return this.wordToLineNumbersMap;
 			}
 
-			// For every word in every line, map its ranges for fast lookup
-			for (let lineNumber = 1, len = model.getLineCount(); lineNumber <= len; ++lineNumber) {
-				const lineContent = model.getLineContent(lineNumber);
-
-				// If line is too long then skip the line
-				if (lineContent.length > MAX_TOKENIZATION_LINE_LEN) {
-					continue;
-				}
-
-				model.forceTokenization(lineNumber);
-				const lineTokens = model.getLineTokens(lineNumber);
-				for (let tokenIndex = 0, tokenCount = lineTokens.getCount(); tokenIndex < tokenCount; tokenIndex++) {
-					const tokenStartOffset = lineTokens.getStartOffset(tokenIndex);
-					const tokenEndOffset = lineTokens.getEndOffset(tokenIndex);
-					const tokenType = lineTokens.getStandardTokenType(tokenIndex);
-					const tokenStr = lineContent.substring(tokenStartOffset, tokenEndOffset);
-
-					// Token is a word and not a comment
-					if (tokenType === StandardTokenType.Other) {
-						DEFAULT_WORD_REGEXP.lastIndex = 0; // We assume tokens will usually map 1:1 to words if they match
-						const wordMatch = DEFAULT_WORD_REGEXP.exec(tokenStr);
-
-						if (wordMatch) {
-							const word = wordMatch[0];
-							if (!this.wordToLineNumbersMap.has(word)) {
-								this.wordToLineNumbersMap.set(word, []);
-							}
+			this.tokenizeViewPort(model);
 
-							this.wordToLineNumbersMap.get(word)!.push(new Position(lineNumber, tokenStartOffset));
-						}
+			this.editor.onDidScrollChange(() => {
+				this.editor.getVisibleRanges().forEach(visibleRange => {
+					if (model.isCheapToTokenize(visibleRange.endLineNumber)) {
```

Comment:
This logic seems flawd to me. So if a end of the range is cheap to tokenize we will tokenize the whole document?!
IMHO we need a check to see if the whole model is cheap to tokenize, and if yes we tokenize the whole thing. If not always tokenize the viewport. @rebornix would know better if something like that exists

Also if you tokenize the complete document there is no longer a need to listen on the scroll change event. I would just dispoe of the listener in that case.

Reply:
I thought about it again and realized that I didn't do a detailed explaination for the tokenization idea, sorry for that.

The reason that we can use `isCheapToTokenize` to reduce the time is there are always tokens generated for the viewport (introduced by Alex's recent changes). Whenever we open a file, scroll to some pages, we will tokenize the viewport based on some heuristics. It doesn't mean they are 100% correct, but almost. So if the viewport is cheap to tokenize, then we force tokenize them. If not, we don't do anything, just read the tokens by `getLineTokens`.
 
`isCheapToTokenize` is `false` in following two major cases

1. if the previous line of a position is not tokenized or invalid (for example, it has dirty changes)
2. if current line is really long

So say you debug into a new file, the editor tokenizes the viewport internally (99.9% correct), and then start a full document tokenization in the background. If now you ask the editor if the viewport is cheap to tokenize or not, it depends on how fast the background tokenization runs.

Since we are only fetching `word` information here, there are fewer chances that the tokens are siganificantly wrong. I'd like to propose

* Check if the `visibleRange.startLineNumber` is easy to tokenize or not
* If so, tokenize the viewport only. (We don't need to `tokenizeCompleteDocument` as there is already a background job for that)
* If not, just read the tokens, they are very unlikely to be wrong, for word information